### ISSUE TITLE
Question: R2RML on Spark

### ISSUE THREAD

**Issue by jindrichmynarz** [SANSA-Stack/SANSA-Stack#213]
Hi,

I'm piecing together the available documentation about Sansa's support for [R2RML](https://www.w3.org/TR/r2rml/) on Spark. If I understand it correctly, [Querying heterogeneous data sources (data lake)](https://github.com/SANSA-Stack/SANSA-Stack/blob/develop/sansa-query/sansa-query-spark/src/main/scala/net/sansa_stack/query/spark/datalake/README.MD) is the most relevant documentation. The method requires 3 arguments: `query`, `mappingsFile`, and `configFile`.

If I read [this](https://github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-datalake#3-sparql-query-interface) right, `query` is limited to SPARQL `SELECT` queries. Other query forms, such as `CONSTRUCT`, are not supported. In that case I assume exporting the entire database to RDF would have to be done via `SELECT ?s ?p ?o WHERE { ?s ?p ?o . }` and transforming the resulting Spark `DataFrame` into [RDF/N-Triples](https://www.w3.org/TR/n-triples/) or another RDF syntax?

The `mappingsFile` is actually in [RML](https://rml.io/specs/rml/), instead of R2RML. I'm wondering what part of the R2RML standard is implemented by Sansa. I noticed [here](https://github.com/SANSA-Stack/SANSA-Stack/blob/159f59caa165ef789bc3ec93e28762b786291dba/sansa-query/README.md) that Sansa wraps [Ontop](https://ontop-vkg.org/) for R2RML. I would expect then that the compliance with R2RML is the same as reported for Ontop in the [R2RML Implementation Report](https://kg-construct.github.io/r2rml-implementation-report/). Is that true?

The `configFile` configures [data sources](https://github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-datalake#2-data-connection-configurations). I see that relational databases for R2RML are supported via JDBC connections. However, the [available connectors](https://github.com/SANSA-Stack/SANSA-Stack/tree/develop/sansa-datalake#connectors) don't include any relational databases. Is this list of connectors up-to-date? Is there a schema or documentation of the expected structure of the `configFile`? The best I found is the [example](https://github.com/SANSA-Stack/SANSA-Stack/blob/159f59caa165ef789bc3ec93e28762b786291dba/sansa-examples/sansa-examples-spark/src/main/resources/datalake/config). The example configuration sets up a database table via the `dbtable` option. How it should be used when mapping an entire database? Should each mapped database table be configured as a separate JDBC source?

Please let me know if there's any better documentation for R2RML on Spark via Sansa, and if my understanding of the current documentation is correct.

**Comment by Aklakan** [comment]
Hi Jindrich, 

- Datalake is a no longer maintained research prototype. For the moment it is probably academic-wise most relevant to your question.
- R2RML is used internally by the vertical partitioner (VP) which creates one table per tuple of (subject-term-type, predicate, object termtype/datatype): The VP generates R2RML documents for these tables which can then be passed to Sparqlify and/or Ontop (or any other compliant engine). So its used internally as a uniform model to configure engines. Conversely, programmatically, the infrastructure to set up a custom spark database layout and then configure a system with a corresponding R2RML document exists and is being used.
- That said, vertical partitioning so far has not been proven to be an efficient way for general purpose query answering.

- Sansa has another engine which so far I simply referred to as the "Binding Engine" (still need to make up a name for it)- it's essentially a port of Jena's query engine which createsRDD[Binding] instances rather than single threaded Iterator[Binding]. 
An this one we already used to create a quite competitive version of Tarql - from the [documentation](http://sansa-stack.github.io/SANSA-Stack/cli/tarql.html):
![image](https://user-images.githubusercontent.com/839608/222561474-5b64d93b-8c98-426b-8141-5f1fc15a096a.png)

Sansa tarql was part of our [latest release](https://github.com/SANSA-Stack/SANSA-Stack/releases/tag/sansa-0.8.0-rc3) With our recent work we extended this approach to RML/CSV. It will release soon and should - from the practical side - supersede datalake.

> I would expect then that the compliance with R2RML is the same as reported for Ontop in the [R2RML Implementation Report](https://kg-construct.github.io/r2rml-implementation-report/). Is that true?

I would assume that to be the case - @LorenzBuehmann has integrated the compliance suites exactly for the purpose of answering that question - probably its easier for him to answer this question :)
